### PR TITLE
Adds service callback with service context

### PR DESCRIPTION
**Pull request by norro**
Adds an additional callback type and handling that allows passing
additional service context (void *) to a service callback.

https://github.com/ros2/rclc/issues/35

### CODE REVIEW

**Comment by norro** [comment]
@JanStaschulat can you have a look at the failing tests, e.g., [in rolling](https://build.ros2.org/job/Rpr__rclc__ubuntu_focal_amd64/60/)?
1. I don't get the reason for the non-equalities, maybe you can give me a hint
1. I have different tests failing in my local rolling setup. This might indicate, that these tests are generally unstable?

**Comment by JanStaschulat** [comment]
@norro Please create a pull request to micro-ros/rclc. The default configuration is ros2/rclc, therefore you always have to change the repository to which you want to create a pull request to. 

I looked briefly at the output errors. These are all from the th OSRF/OpenRobotics build farm and not the github action build job. I know that some test cases of the rclc executor fail sometimes due to timing issues at the OSRF build farm. I don't have a solution for that yet. Sometimes they get through sometimes not. 

For the Eloquent build job, there seems to be a problem with the docker container, as 'ament_cmake_ros' package is not found. We will remove Eloquent from our tests, as Eloquent will not be supported any more.

File: rclc/src/rclc/executor.c [hunk 1]
```diff
@@ -430,6 +430,59 @@ rclc_executor_add_service(
...
+
+  // increase index of handle array
+  executor->index++;
+
+  // invalidate wait_set so that in next spin_some() call the
+  // 'executor->wait_set' is updated accordingly
+  if (rcl_wait_set_is_valid(&executor->wait_set)) {
+    ret = rcl_wait_set_fini(&executor->wait_set);
+    if (RCL_RET_OK != ret) {
+      RCL_SET_ERROR_MSG("Could not reset wait_set in rclc_executor_add_service function.");
```
Comment (JanStaschulat) [inline rclc/src/rclc/executor.c:476]:
```suggestion
      RCL_SET_ERROR_MSG("Could not reset wait_set in rclc_executor_add_service_with_context function.");
```

File: rclc/src/rclc/executor.c [hunk 2]
```diff
@@ -430,6 +430,59 @@ rclc_executor_add_service(
...
+      return ret;
+    }
+  }
+
+  executor->info.number_of_services++;
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Added a service.");
```
Comment (JanStaschulat) [inline rclc/src/rclc/executor.c:482]:
```suggestion
  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Added a service_with_context.");
```

**Review by JanStaschulat — CHANGES_REQUESTED** [review]
Only minor improvements.

**Comment by JanStaschulat** [comment]
> @JanStaschulat can you have a look at the failing tests, e.g., [in rolling](https://build.ros2.org/job/Rpr__rclc__ubuntu_focal_amd64/60/)?
> 
>     1. I don't get the reason for the non-equalities, maybe you can give me a hint
> 
>     2. I have different tests failing in my local rolling setup. This might indicate, that these tests are generally unstable?

I know that sporadically these units tests fail. This is a timing issue. For now, just ignore these error from trigger_one trigger_always trigger_once etc. In next sprint (February) I will try to re-write them. On my native Linux and at the CI github action on both micro-ros and ros2/rclc they always work. Only at the build farm of OSRF they fail. It is interesting that they also fail at your local machine. Could you please list the unit test, that fail?

**Comment by norro** [comment]
Created a PR in micro-ros/rclc against micro-ros/rclc master: https://github.com/micro-ROS/rclc/pull/68

**Pull request closed without merging**